### PR TITLE
compute: log timestamps from recursive queries

### LINKED ISSUE #21725: Reachability logging does not cover recursive queries
Reachability logging tries to downcast timestamps to `Timestamp`. For recursive queries, the type is different: It is a product time, which internally is a vector of `Timestamp`. We need to consider this case to surface full reachability data.

### PR DESCRIPTION
**Pull request by sdht0**
Reachability logging currently logs a `null` Timestamp for recursive queries with multi-dimensional timestamps. This PR changes that to always log the first dimension. 

Fixes #21725

### CODE REVIEW

File: src/compute/src/logging/initialize.rs [hunk 1]
```diff
@@ -10,24 +10,26 @@ use std::collections::BTreeMap;
 use std::rc::Rc;
 use std::time::{Duration, Instant};
 
+use crate::arrangement::manager::TraceBundle;
+use crate::extensions::arrange::{KeyCollection, MzArrange};
+use crate::logging::compute::ComputeEvent;
+use crate::logging::{BatchLogger, EventQueue, SharedLoggingState};
```
Comment (antiguru) [inline src/compute/src/logging/initialize.rs:16]: Nit: Move down to after imports from other crates.

Our style is:
* std imports
* other crate imports
* this crate imports
* Separated by a newline.

File: src/compute/src/logging/initialize.rs [hunk 2]
```diff
@@ -230,3 +232,17 @@ impl<A: Allocate + 'static> LoggingContext<'_, A> {
         )
     }
 }
+
+#[inline]
+fn downcast_timestamp(time: &dyn ProgressEventTimestamp) -> Option<Timestamp> {
```
Comment (antiguru) [inline src/compute/src/logging/initialize.rs:237]: Maybe call the function `try_downcast_timestamp`? Also, add documentation.

**Review by antiguru — APPROVED** [review]
LGTM, left some small comments.

File: src/compute/src/logging/initialize.rs (hunk 2 above)
Comment (sdht0) [reply src/compute/src/logging/initialize.rs]: Done. Added some documentation, is that sufficient?

**Review by antiguru — APPROVED** [review]

**Pull request merged**